src/ibmca_ec.c

```diff
@@ -0,0 +1,918 @@
+    if (rc != 0) {
+        IBMCAerr(IBMCA_F_ICA_EC_KEY_GET_PRIVATE_KEY, rc);
+        goto end;
+    }
+
+    /* Add private key data to EC_KEY */
+    privkey = BN_bin2bn((unsigned char*)&d, d_len, NULL);
+    if (!EC_KEY_set_private_key(eckey, privkey)) {
+        IBMCAerr(IBMCA_F_IBMCA_EC_KEY_GEN, IBMCA_R_EC_INTERNAL_ERROR);
+        goto end;
+    }
+
+    ret = 1;
+
+end:
+    p_ica_ec_key_free(icakey);
+    EC_POINT_free(pubkey);
+    BN_clear_free(privkey);
+    BN_clear_free(bn_x);
+    BN_clear_free(bn_y);
+
+    return ret;
+}
+ #endif
+
+/**
+ * ECDH key derivation method, replaces ossl_ecdh_compute_key.
+ *
+ * @return 1 success
+ *         0 error
+ */
+int ibmca_ecdh_compute_key(unsigned char **pout, size_t *poutlen,
+                           const EC_POINT *pub_key, const EC_KEY *ecdh)
+{
+    ICA_EC_KEY *ica_pubkey = NULL, *ica_privkey = NULL;
+    const EC_GROUP *group;
+    BIGNUM *bn_d, *bn_x, *bn_y;
+    unsigned int n, i, privlen;
+    unsigned char X[IBMCA_EC_MAX_D_LEN];
+    unsigned char Y[IBMCA_EC_MAX_D_LEN];
+    unsigned char D[IBMCA_EC_MAX_D_LEN];
+    unsigned char *z_buf = NULL;
+    int rc, ret = 0, nid;
+ #ifndef OLDER_OPENSSL
+    int (*compute_key_sw)(unsigned char **pout, size_t *poutlen,
+                          const EC_POINT *pub_key, const EC_KEY *ecdh) = NULL;
+ #endif
+
+    /* Get group from EC_KEY */
+    if ((group = EC_KEY_get0_group(ecdh)) == NULL) {
+        IBMCAerr(IBMCA_F_IBMCA_ECDH_COMPUTE_KEY, IBMCA_R_EC_INVALID_PARM);
+        return 0;
+    }
+
+    /* Determine curve nid */
+    nid = EC_GROUP_get_curve_name(group);
+    if (nid <= 0) {
+        IBMCAerr(IBMCA_F_IBMCA_ECDH_COMPUTE_KEY, IBMCA_R_EC_INTERNAL_ERROR);
+        return 0;
+    }
+
+    /* Create ICA_EC_KEY object for public key */
+    ica_pubkey = p_ica_ec_key_new(nid, &privlen);
+    if (ica_pubkey == NULL) {
+        /* This curve is not supported by libica. */
+ #ifdef OLDER_OPENSSL
+        return 0;
+ #else
+        EC_KEY_METHOD_get_compute_key(ossl_ec, &compute_key_sw);
+        if (compute_key_sw == NULL) {
+            IBMCAerr(IBMCA_F_IBMCA_ECDH_COMPUTE_KEY,
+                     IBMCA_R_EC_INTERNAL_ERROR);
+            return 0;
+        }
+
+        return compute_key_sw(pout, poutlen, pub_key, ecdh);
+ #endif
+    }
+
+    /* Get (X,Y) from EC_POINT */
+    bn_x = BN_new();
+    bn_y = BN_new();
+    if (!EC_POINT_get_affine_coordinates_GFp(group, pub_key, bn_x, bn_y, NULL)) {
+        IBMCAerr(IBMCA_F_IBMCA_ECDH_COMPUTE_KEY, IBMCA_R_EC_INTERNAL_ERROR);
+        goto end;
+    }
+
+    /* Format (X) as char array, with leading zeros if necessary */
+    n = privlen - BN_num_bytes(bn_x);
+    for (i = 0; i < n; i++)
+        X[i] = 0x00;
+    BN_bn2bin(bn_x, &(X[n]));
+
+    /* Format (Y) as char array, with leading zeros if necessary */
+    n = privlen - BN_num_bytes(bn_y);
+    for (i = 0; i < n; i++)
+        Y[i] = 0x00;
+    BN_bn2bin(bn_y, &(Y[n]));
+
```

Comment:
and here the very same. Maybe a macro or a static function which does this format is a better way.